### ISSUE TITLE
Support dynamic configuration

### ISSUE THREAD

**Issue by jolestar** [fluent/fluent-bit#365]
Please support reload configuration files and do not need restart server. 
I think there are two approach:

1.  Catch SIGHUP signal
2. Offer a http api

**Comment by jasonkeene** [comment]
SIGHUP currently takes down the process:

```
Fluent-Bit v0.13.4
Copyright (C) Treasure Data

[engine] caught signal (SIGHUP)
```

It would be great if this reloaded the config without taking down the process.

**Comment by nikolay** [comment]
It would be nice to watch the configuration files as well and support hot configuration reloading as well.

**Comment by vinkdong** [comment]
If there is an API like prometheus `/-/reload` , it’s even better.

**Comment by discordianfish** [comment]
Or just use inotify to watch the confiug file and automatically reload. It's using inotify for the log anyway, right?

**Comment by jasonkeene** [comment]
Not that I would do this, but I could imagine users modifying their config file on the server and writing out progress. If you reload on writes like that it may load an invalid or partial config. Also, this breaks from most server software that receive an explicit signal like `SIGHUP`.

**Comment by edsiper** [comment]
What's the current problem of restarting Fluent Bit using systemd ?, I would like to learn the real need of config reload without restarting, as far as I can tell services that offer this always start a new process (behind the scenes).

The real complexity will be to add this handling on every single input plugin. Please elaborate the use-case.

**Comment by discordianfish** [comment]
@edsiper I'm running fluent-bit in a container on kubernetes, so it's not using systemd. Running another process manager in the container, at least on kubernetes, is an antipattern since you loose the process management features provided by kubernetes (restart/policies, monitoring, event log etc).

That being said, there is a case to be made for it being kubernetes's job to restart the fluent-bit container when the config changes. Something that's being discussed in https://github.com/kubernetes/kubernetes/issues/22368

Currently the best option would be reloading on:

- inotify (people would need to write the config update atomically, but this is an established pattern and can be easily achieved by `mv` the file to the final location after creating it somewhere else)
- @vinkdong's suggestion of an http endpoint which can be used with https://github.com/jimmidyson/configmap-reload as a sidecar container (it would watch the config file via inotify and send the http request to reload)

Reloading on SIGHUP should also work but requires an alpha feature: https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/

**Comment by nikolay** [comment]
@discordianfish The answer @edsiper speaks volumes! I had so many problems with Fluent Bit that I switched to Filebeat, which is mature, written in Go so more people can contribute, and it has most of the features Fluent Bit is not even planning to have any time soon, including configuration reloading.

**Comment by donbowman** [comment]
I think adding an inotify on the config and restart would be good.

the next step would be to emulate grafana and have multiple config maps, one for each pod that wants to have custom log message formats. This means that fluent-bit watches for new configmaps to be written, and reads them as they happen. (see https://github.com/coreos/prometheus-operator).

e.g. in my pod i could have a 'fluent-parser' and have a single parser entry, which would be dynamically loaded.

the restart can be ok as long as the trackingoffsets is used. Otherwise its disasterous.

**Comment by edsiper** [comment]
I am not against configuration reloading, but understanding the __why__ and it urgency helps to assign a priority, as you can guess with different interfaces in input, filters and outputs is not a straightforward change. As said above, for Kubernetes use-case config-reloading is an anti-pattern (immutable).

Now about Fluent Bit v/s other projects, I consider that is really good to learn from others, that's the way to go. We still have a plenty of room for improvements.

About programming languages, consensus is something that will unlikely exists and having different preferences is just fine. Users decide their tooling, hopefully they have choices, Fluent Bit is one of them.

**Comment by nikolay** [comment]
@donbowman I don't think C, C++, Rust, and Go have much in common outside of being compiled languages.

**Comment by donbowman** [comment]
they are 'system-level' languages.
C++ definitely has more features, larger standard libraries, and more
programmers that use it than the others (except for C which is the clear
winner in usage).

https://stackify.com/popular-programming-languages-2018/

its not a popularity contest tho, and the languages don't give any inherent
capabilities to the product.

go and C have the lowest level of abstraction of the 4 (e.g. no classes, no
exceptions).

**Comment by nikolay** [comment]
@donbowman There's a huge legacy codebase on C and C++ and this drives its "popularity". The Kubernetes ecosystem, which I personally care about, is dominated by Go. Here's a nice analysis of the subject by a very popular project, which analyzed the options in 2018 and on: https://blog.phusion.nl/2018/09/18/migrating-passenger-from-cxx-to-go/

**Comment by altenhof** [comment]
Coming back to the original question, I'm wondering how I could use fluent-bit to support the following use case:

Customer is running a k8s cluster and wants to ship logs to an external "service". Service provisioning and binding follows the OSB API pattern, i.e. customer creates a service instance and then binds an "app" to that instance. 

The result of the binding operation is a secret that fluent-bit would then (have to) digest to connect to the remote service instance.

One could, of course (like always) put an additional layer between fluent-bit and the remote service endpoint, e.g. have fluent-bit send to a local service proxy and make that service proxy listen to/watch the secret, but that doesn't sound like a good solution.

Any other opinions/suggestions?

**Comment by donbowman** [comment]
I have an outstanding PR to fluent-bit https://github.com/fluent/fluent-bit/pull/842 that should solve this [original] issue.

is the secret ok to have in your configmap for fluent-bit? In which case my PR achieves your goal.
Or is it something else that fluent-bit doesn't yet support?

where does the secret get injected, what is the output driver in use?

**Comment by altenhof** [comment]
@donbowman,

> s the secret ok to have in your configmap for fluent-bit?
Yes
> where does the secret get injected, what is the output driver in use?
Not sure whether I fully get this. 

AFACT, the workflow that I described is what [Service Catalog](https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/) supposed to do. 

The Docker output driver is json-file. Or did you mean something else?

**Comment by donbowman** [comment]
i mean the fluent-bit output driver. e.g. how does the secret get injected.

**Comment by altenhof** [comment]
Currently, it's `out_http`.

**Comment by donbowman** [comment]
so presumably you end up with an auth-token that goes in a header, and want to use the `Header` feature of out_http with this JWT?

so some agent registers, gets the JWT, updates the fluent-bit config map, and fluent-bit auto-reloads that?

**Comment by altenhof** [comment]
JWT auth token in `Header` feature of out_http? Where is that documented? Can't find it in https://fluentbit.io/documentation/current/output/http.html.

All I see there is `HTTP_User` and `HTTP_Password`. 

Anyways, yes, it seems like there needs to be some sort of agent running, read auth data from secret, update config map, etc.

Or, sigh, have fluent-bit talk to local proxy and let the local proxy do that magic.

**Comment by donbowman** [comment]
https://docs.fluentbit.io/manual/output/http

I was thinking the Header field, you can add the auth-header/jwt tag as needed there.
but you'd have to get it into the configmap.

**Comment by edsiper** [comment]
I've changed the title of the issue since due to the feedback received internally and on GitHub what we need is dynamic configuration mechanisms which are more than a "config file reload". 

After 1.0 we will focus on the design of this feature which involves having a parent process to monitor the engine plus other extensible interfaces to make this happens.

**Comment by altenhof** [comment]
Hi @edsiper,
thanks for the update. Just out of curiosity: What does "after 1.0" mean in terms of dates?

**Comment by jiaxuanzhou** [comment]
hi @edsiper 
is there any update?

**Comment by kpbll** [comment]
I am waitng for this because we use forward(tcp) input plugin in in-memory buffering, but want to store data in daily-based indexes in elasticsearch. Have no ideas how to implement that (

**Comment by jroper** [comment]
Use case here: In Kubernetes, we configure fluentbit using a configmap. We have a multitenant platform where tenants can configure their own log outputs, whenever they change their log output configuration we update the config map, and then we need fluentbit to pick up this configuration somehow. Currently, the only way to do this is to update a dummy environment variable in the fluentbit `DaemonSet`, to trigger a full redeploy. It would be nice if fluentbit could pick up the change without a full redeploy. It doesn't even have to dynamically reconfigure itself, just watch the files, and if they change, exit. Kubernetes will restart the process.

**Comment by MOZGIII** [comment]
Since k8s 1.15 you can use `kubectl rollout restart`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#-em-restart-em-) to recreate pods so they pick up new configuration.
Much better than editing a dummy env var. 😄

**Comment by donbowman** [comment]
we moved to use kustomize, it hashes the config-map name based on contents, thus achieving the rolling restart when changed.

**Comment by dharmab** [comment]
Use case: We use Fluent Bit as a sidecar container in Kubernetes. Currently, making a minor change to our logging config requires us to restart all applications that use the sidecar.

**Comment by mcdano** [comment]
Does Fluent Bit re-read Kubernetes Secrets at runtime?  

Scenario: sending logs with tcp output and our ssl cert expires.  We need to update the secret with the new cert.  Will Fluent Bit pickup of the new secret during runtime without the need for a pod redeploy? Or is a redeploy required?

**Comment by edsiper** [comment]
re-deploy is required. How often your SSL cert expires ?

**Comment by mcdano** [comment]
> re-deploy is required. How often your SSL cert expires ?

Approx every 6 months.

**Comment by donbowman** [comment]
a rollout restart deployment is sufficient.
most people have 90 expiry, refreshed in 45-50 days, due to use of let's encrypt.

**Comment by sbaier1** [comment]
for the k8s sidecar use-case, especially when used with stateful applications that have a high cost for a rolling update, this is honestly not really an acceptable solution. IMO fluent-bit should be capable of detecting changes to the config file at run-time in some way, or at least provide an endpoint to update the config

**Comment by lukasmrtvy** [comment]
Exactly, its nearly impossible to manager rollout restart in bigger scales, of course, You can always use https://github.com/stakater/Reloader, https://github.com/pusher/wave, or similar, but that should not be necessary.

**Comment by FeynmanZhou** [comment]
Take a look at the Fluentbit-operator: https://github.com/kubesphere/fluentbit-operator

Fluentbit-operator addressed this issue and facilitates the deployment of Fluent Bit and provides great flexibility in building a logging layer based on Fluent Bit. For now, it provides the following features:

- Fluent Bit Management: Deploy and destroy Fluent Bit Daemonset automatically.
- Custom Configuration: Select input/filter/output plugins via labels.
- Dynamic Reloading: Update configuration without rebooting Fluent Bit pods.

**Comment by tpetrov-lp** [comment]
@edsiper I want to share another use case for configuration changes scan. 
Imagine multiple teams working on different services. Each service should be deployed independently from potentially different team. Let's say that a service has some log filtering (grep like) requirements which the owner (some team) defines. Upon service deploy, the filter should be updated and ideally this should not cause rollout restart. It may happen multiple times per day to deploy a service. (of course changing the grep logic is less frequent)

That's a good candidate for configuration changes scan, unless you propose a better solution.

**Comment by noniu** [comment]
is there any update?

**Comment by stevehipwell** [comment]
@edsiper @patrick-stephens is there any progress on this? If monitoring the config file is too expensive I'd be interested if a reload endpoint would be possible?

**Comment by GRomR1** [comment]
it was implemented with release [2.1.0](https://fluentbit.io/announcements/v2.1.0/), isn't it?

https://docs.fluentbit.io/manual/administration/hot-reload

would be great to add this in [logging-operator](https://github.com/kube-logging/logging-operator) project
